README.md
# TradingAPIs

Mega-Repository holding many API wrappers I have created to access trading data and/or functionality.
Mainly leverages servant-client package for Haskell.


APIOanda/ChangeLog.md
# Changelog for APIOanda

## Unreleased changes


APIPolygon/app/Main.hs
-- stack --resolver lts-13.7 script

{-# LANGUAGE DataKinds         #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators     #-}

import System.IO                (IOMode (ReadMode), openFile, hGetLine)
import Data.Text                (pack)
import Control.Monad.IO.Class   (liftIO)
import Network.HTTP.Client      (newManager)
import Network.HTTP.Client.TLS  (tlsManagerSettings)
import Servant.Client           (mkClientEnv, runClientM, parseBaseUrl)

import System.Environment ()
-- for handling array of results
import qualified Data.Vector as V

import API.PolygonApi


-- https://api.polygon.io/v2/aggs/ticker/X:ADAUSD/range/1/day/2021-07-24/2021-07-27?apiKey=O5ZnLHNhwqPLC_EkdspvZxFPGRPhxtI8

-- maybe make it auto refresh/get new key automatically
keyPath =  "../../APIkeys/PolygonAPI.txt"

-- TODO: make it so can feed key at any time
main :: IO ()
main = do
  key <- hGetLine =<< openFile keyPath ReadMode
  manager' <- newManager tlsManagerSettings
  liftIO . print =<< (run manager' $ aggTrades (pack "X:ADAUSD") 1 (pack "day") (pack "2021-07-24") (pack "2021-07-27") (pack key))
  liftIO . print =<< (run manager' $ prevOHLC (pack "X:ADAUSD") (pack key))
                    

APIPolygon/src/API/Types.hs
{-# LANGUAGE DataKinds         #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators     #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}


module API.Types
    ( endpoint
    , AggTrades
    , Ticker
    , Multiplier
    , Timespan
    , FromDate
    , ToDate 
    , ApiKey
    ) where

import qualified Data.Text  as T
import API.Prelude


type Ticker = T.Text
type Multiplier = Int
type Timespan = T.Text
type FromDate = T.Text
type ToDate = T.Text
type ApiKey = T.Text

endpoint :: T.Text
endpoint = "api.polygon.io"

data OHLCbar = OHLCbar 
  { v :: Double     -- volume
  , vw :: Double    -- VWAP
  , o :: Double     -- open
  , c :: Double     -- close
  , h :: Double     -- highest
  , l :: Double     -- lowest
  , t :: Integer    -- start time in Unix Msec
  , n :: Int        -- number of trades
  }
  deriving (Show, Generic, FromJSON)

type OHLCbars = [OHLCbar]

data AggTrades = AggTrades 
  { status :: String
  , ticker :: String
  , queryCount :: Int
  , resultsCount :: Int -- should equal querycount for complete result
  , adjusted :: Bool
  , results :: OHLCbars
  }
  deriving (Show, Generic, FromJSON)


APIPolygon/src/API/Prelude.hs
{-# LANGUAGE DataKinds         #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators     #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}

module API.Prelude
    ( module X
    ) where

import Control.Exception                        as X
    (throw)
import Data.Text                                as X
    (pack, unpack)
import GHC.Generics                             as X 
    ( Generic )
import Data.Aeson                               as X
    ( FromJSON
    , defaultOptions, genericParseJSON
    , fieldLabelModifier, parseJSON
    , withObject, (.:)
    )
import Data.Proxy                               as X
    ( Proxy(..) )
import Servant.API                              as X
    ( JSON, QueryParam, Get, Capture
    , type (:>), type (:<|>)((:<|>)))
import Servant.Client                           as X
    ( ClientM, ClientError
    , client, mkClientEnv, runClientM
    , parseBaseUrl
    )
import Network.HTTP.Client                      as X
    (Manager)




APIPolygon/src/API/PolygonApi.hs
{-# LANGUAGE DataKinds         #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators     #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}

module API.PolygonApi
    ( run
    , aggTrades
    , prevOHLC
    , queries
    ) where


import API.Prelude
import qualified API.Types as AT

type PolygonAPI =
  ("v2":>
    ("aggs" :> ("ticker" :> Capture "ticker" AT.Ticker
            -- aggregate OHLC Bars
                         :> "range" :> Capture "multiplier" AT.Multiplier
                         :> Capture "timespan" AT.Timespan
                         :> Capture "from" AT.FromDate
                         :> Capture "to" AT.ToDate
                         -- :> QueryParam "adjusted" AdjustSplit
                         -- :> QueryParam "sort" WillSort
                         -- :> QueryParam "limit" QueryLimit
                         :> QueryParam "apiKey" AT.ApiKey
                         :> Get '[JSON] AT.AggTrades
                    -- previous OHLC
           :<|> "ticker" :> Capture "ticker" AT.Ticker
                         :> "prev" :> QueryParam "apiKey" AT.ApiKey
                         :> Get '[JSON] AT.AggTrades)))

polygonApi :: Proxy PolygonAPI
polygonApi = Proxy

-- queries to each endpoint of api
aggTrades' :: AT.Ticker -> AT.Multiplier
          -> AT.Timespan -> AT.FromDate -> AT.ToDate 
          -> Maybe AT.ApiKey -> ClientM AT.AggTrades
prevOHLC' :: AT.Ticker -> Maybe AT.ApiKey -> ClientM AT.AggTrades
-- generate functions given the queries
aggTrades' :<|> prevOHLC' = client polygonApi


-- apiCalls which map to the above queries
aggTrades :: AT.Ticker -> AT.Multiplier -> AT.Timespan
           -> AT.FromDate -> AT.ToDate -> AT.ApiKey -> ClientM AT.AggTrades
aggTrades ticker multiplier timespan from to apiKey =
  aggTrades' ticker multiplier timespan from to (Just apiKey)

prevOHLC :: AT.Ticker -> AT.ApiKey -> ClientM AT.AggTrades
prevOHLC ticker apiKey = prevOHLC' ticker (Just apiKey)

queries :: AT.Ticker -> AT.Multiplier -> AT.Timespan -> AT.FromDate -> AT.ToDate 
        -> AT.ApiKey -> ClientM (AT.AggTrades, AT.AggTrades)
queries ticker multiplier timespan from to apiKey = do
  agg <- aggTrades' ticker multiplier timespan from to (Just apiKey)
  prev <- prevOHLC' ticker (Just apiKey)
  return (agg, prev)

{- calling the apis-}
-- runClientM, but with the ClientEnv already given
type Runner a = ClientM a -> IO a

-- TODO: have in readerT format to get the Manager
-- code: asks config
-- run :: Manager -> ClientM a -> IO (Either ClientError a)
-- run mgr apiCall = either throw return =<<
--   runClientM apiCall (mkClientEnv mgr apiPath)
--   where
--     apiPath = parseBaseUrl (unpack AT.endpoint)

run :: Manager -> ClientM a -> IO (Either ClientError a)
run mgr apiCall = do
  apiPath <- parseBaseUrl (unpack AT.endpoint)
  -- runClientM apiCall (mkClientEnv mgr apiPath) >>= return
  runClientM apiCall (mkClientEnv mgr apiPath)
  
  

{-
main :: IO ()
main = do
  manager' <- newManager tlsManagerSettings
  apiPath <- parseBaseUrl "api.polygon.io"
  res <- runClientM (queries (T.pack "X:ADAUSD") 1 (T.pack "day")
                             (T.pack "2021-07-24") (T.pack "2021-07-27")
                             (T.pack key))
                    (mkClientEnv manager' apiPath)
  -- print res
  case res of
    Left err -> putStrLn $ "Error: " ++ show err
    Right (aggTrades, prevOHLC) -> do 
      print aggTrades
      print prevOHLC
-}